### DISCUSSION TITLE
Trying to autogenerate pojos during compile time in sbt

### DISCUSSION THREAD

**Discussion by nnbbz** [kobylynskyi/graphql-java-codegen#550] [General]
Hello, thank you providing the sbt plugin for the codegen!  I just had a question.  Currently, I have to run sbt graphqlCodegen (to autogenerate the pojos) and then run sbt compile to compile the code.

What I want to do is trigger graphqlCodegen automatically when I run sbt compile.  I've been trying to get this working in my build.sbt but with no luck (perhaps my syntax is not correct?)

I have the following line in my build.sbt, but it complains that graphqlCodegen is not defined.  Can you shed some light on how I can do this?

(compile in Compile) := ((compile in Compile) dependsOn graphqlCodegen).value

I'm new to SBT

**Comment by jxnu-liguobin** [comment]
@nnbbz 
Please use ```(compile in Compile) := ((compile in Compile) dependsOn (graphqlCodegen in GraphQLCodegenConfig)).value```

In SBT plugins, there is a concept of scope to distinguish between multi module and multi project. So you have to specify the task in which scope.

Because  the generated code will be changed which should be known to users. (Although I'm a  Scala developer, I'm not sure it's appropriate, so I didn't do that for the plugin).

If you think it's necessary to build an issue to track it, I'll consider implementing it.

**Comment by nnbbz** [comment]
@jxnu-liguobin 
Thank you, that worked!

Unfortunately, when I do `sbt run`, it looks like it invokes `sbt compile` whenever sources change.  Which will now invoke the plugin and I get this exception:

Caused by: java.lang.Exception: Unable to delete directory
        at io.github.dreamylost.graphql.codegen.GraphQLCodegenPlugin.$anonfun$projectSettings$14(GraphQLCodegenPlugin.scala:225)
        
Is there anyway for the plugin to not delete the directory and instead just delete/replace the files.  Or make that an option?  I understand why you would want to delete the directory (to remove old autogenerated classes)

**Reply by jxnu-liguobin** [reply]
@nnbbz 
Sorry, I think this behavior is SBT's own, not plugin's.
This plugin will only listen to graphql schema which end with ".graphql" or "graphqls". As follows:
![image](https://user-images.githubusercontent.com/29496873/108582697-cf373d00-736f-11eb-92a0-f75280930100.png)

The plugin does not delete the files, but because the files are in the target directory, SBT may delete them when clean task.
(Triggered when sources code were changed).

This is related to the listening mechanism of run task. You may need to redefine the behavior of the run task and filter out the files you generate.

**Reply by jxnu-liguobin** [reply]
![image](https://user-images.githubusercontent.com/29496873/108582916-078b4b00-7371-11eb-84be-6bc45573e019.png)
I looked at it carefully, which is more likely to be the reason. This is the behavior of the library itself.
Exception msg is equals to this:
```java
package com.kobylynskyi.graphql.codegen.model.exception;

/**
 * Exception that indicates error while deleting directory
 *
 * @author kobylynskyi
 */
public class UnableToDeleteDirectoryException extends RuntimeException {

    public UnableToDeleteDirectoryException(Exception e) {
        super("Unable to delete directory", e);
    }

}
```
Throw it here when generate method was invoked.
May there be a concurrency problem leading to deletion failure? @kobylynskyi 
```java
    public static void deleteDir(File dir) {
        if (!dir.exists()) {
            return;
        }
        File[] files = dir.listFiles();
        if (files != null) { //some JVMs return null for empty dirs
            for (File subFile : files) {
                if (subFile.isDirectory()) {
                    deleteDir(subFile);
                } else {
                    try {
                        Files.delete(subFile.toPath());
                    } catch (IOException e) {
                        throw new UnableToDeleteDirectoryException(e);
                    }
                }
            }
        }
        try {
            Files.delete(dir.toPath());
        } catch (IOException e) {
            throw new UnableToDeleteDirectoryException(e);
        }
    }
```

**Reply by kobylynskyi** [reply]
@nnbbz, Could you please paste the part of the stacktrace which shows line of failure in `GraphQLCodegenFileCreator.java`.
We need to understand if the issue is with individual file deletion or folder deletion.
Thanks.

**Reply by nnbbz** [reply]
Here you go:

play.sbt.PlayExceptions$UnexpectedException: Unexpected exception[Exception: Unable to delete directory]
        at play.sbt.run.PlayReload$.$anonfun$taskFailureHandler$1(PlayReload.scala:44)
        at scala.Option.map(Option.scala:230)
        at play.sbt.run.PlayReload$.taskFailureHandler(PlayReload.scala:37)
        at play.sbt.run.PlayReload$.compileFailure(PlayReload.scala:30)
        at play.sbt.run.PlayReload$.$anonfun$compile$3(PlayReload.scala:65)
        at scala.util.Either$LeftProjection.map(Either.scala:573)
        at play.sbt.run.PlayReload$.compile(PlayReload.scala:65)
        at play.sbt.run.PlayRun$.$anonfun$playRunTask$4(PlayRun.scala:78)
        at play.runsupport.Reloader.$anonfun$reload$1(Reloader.scala:496)
        at play.runsupport.Reloader$$anon$1.run(Reloader.scala:57)
Caused by: java.lang.Exception: Unable to delete directory
        at io.github.dreamylost.graphql.codegen.GraphQLCodegenPlugin.$anonfun$projectSettings$14(GraphQLCodegenPlugin.scala:225)
        at scala.Function1.$anonfun$compose$1(Function1.scala:49)
        at sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:62)
        at sbt.std.Transform$$anon$4.work(Transform.scala:68)
        at sbt.Execute.$anonfun$submit$2(Execute.scala:282)
        at sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:23)
        at sbt.Execute.work(Execute.scala:291)
        at sbt.Execute.$anonfun$submit$1(Execute.scala:282)
        at sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:265)
        at sbt.CompletionService$$anon$2.call(CompletionService.scala:64)
[error] stack trace is suppressed; run 'last GraphqlCodegen-main / graphqlCodegen' for the full output
[error] (GraphqlCodegen-main / graphqlCodegen) java.lang.Exception: Unable to delete directory

**Reply by jxnu-liguobin** [reply]
Unfortunately, the stack is caught.
```scala
try {
          val _outputDir = (outputDir in GraphQLCodegenConfig).value
          val _introspectionResult = (graphqlQueryIntrospectionResultPath in GraphQLCodegenConfig).value.orNull
          lazy val instantiateCodegen = (mappingConfig: MappingConfig) => {
            (generatedLanguage in GraphQLCodegenConfig).value match {
              case JAVA =>
                new JavaGraphQLCodegen(getSchemas, _introspectionResult, _outputDir, mappingConfig, mappingConfigSupplier)
              case SCALA =>
                new ScalaGraphQLCodegen(getSchemas, _introspectionResult, _outputDir, mappingConfig, mappingConfigSupplier)
              case _ =>
                throw new LanguageNotSupportedException((generatedLanguage in GraphQLCodegenConfig).value)
            }
          }
          result = instantiateCodegen(getMappingConfig().value).generate.asScala
          for (file ← result) {
            sLog.value.success(s"${file.getName}")
          }
        } catch {
          case e: Exception ⇒
            throw new Exception(s"${e.getLocalizedMessage}")
        }
```

**Reply by nnbbz** [reply]
Are you asking me to run that code somewhere? or is that the snippet of how it's being caught?  If you want me to run it, let me know what I can specify in the build.sbt

**Reply by jxnu-liguobin** [reply]
This is the code of the plugin. I'm going to use the play project to reproduce it

**Reply by jxnu-liguobin** [reply]
Do you put the file under the ```app```? Or in the ```target``` directory?
Can you describe the specific process? In the play project, I execute SBT run without repeating the error.

My files are generated correctly, even after changing the source code and reload.
![image](https://user-images.githubusercontent.com/29496873/108585200-1fb79600-7382-11eb-9ec3-45c767151f99.png)

**Reply by nnbbz** [reply]
Interesting, I put the `src_managed` folder next to the `app` directory (not in the target directory)
How did you put it in the `target` directory?  Can you share your settings?  When you put it in the `target` directory, do you have non-generated classes that reference the auto generated classes and does it compile?

![image](https://user-images.githubusercontent.com/76269494/108585722-ca867080-72fe-11eb-8a31-5d0eb3c332c5.png)

**Reply by nnbbz** [reply]
I also did something like this in my `built.sbt`

```
sourceGenerators in Compile += graphqlCodegen in GraphQLCodegenConfig
unmanagedSourceDirectories in Compile := List(baseDirectory.value / "app")
managedSourceDirectories in Compile := List(baseDirectory.value / "auto-generated-src")
```

Read that this would also run graphqlCodegen during compile time.  But gives the same error.  Perhaps if I generate the java classes in the target folder, it will work.  Please let me know how you were able to generate it there.

I have this as my outputDir:

`outputDir in GraphQLCodegenConfig := new File("auto-generated-src"),
`

**Reply by jxnu-liguobin** [reply]
Please take a look. https://github.com/kobylynskyi/graphql-java-codegen/issues/551.

**Reply by jxnu-liguobin** [reply]
> When you put it in the `target` directory, do you have non-generated classes that reference the auto generated classes and does it compile?

Yes, ` app/` can reference auto generated classes which have saved to `generateCodegenTargetPath`.

**Reply by jxnu-liguobin** [reply]
The important ones are `outputDir` and `generateCodegenTargetPath `. In addition, you don't need any config such as `sourceGenerators `, `unmanagedSourceDirectories `, `managedSourceDirectories `.


There are some example projects.They cover almost all possible scenarios for using SBT plugin:

1. In sbt project and save to src/main/java  https://github.com/growingio/growingio-graphql-javasdk/blob/master/build.sbt
![image](https://user-images.githubusercontent.com/29496873/108587042-71651e00-738c-11eb-8075-cd5c45f51f22.png)


2. In sbt project   https://github.com/kobylynskyi/graphql-java-codegen/blob/master/plugins/sbt/graphql-java-codegen-sbt-plugin/src/sbt-test/graphql-codegen-sbt-plugin/example-client-scala/build.sbt
![image](https://user-images.githubusercontent.com/29496873/108587069-88a40b80-738c-11eb-8431-d2395f012a37.png)

3. In play project    https://github.com/jxnu-liguobin/scala-examples/blob/master/build.sbt    
![image](https://user-images.githubusercontent.com/29496873/108587098-a3768000-738c-11eb-9d9e-6fb49b3fae10.png)

**Reply by nnbbz** [reply]
Awesome thank you so much, I will give this a try!

**Reply by nnbbz** [reply]
Hello @jxnu-liguobin 

The above seemed to work for me in the sense that it is now putting the files in the target directory and most of the time it's building fine.  However, sometimes I still get the exception where I cannot delete the directory.

Caused by: java.lang.Exception: Unable to delete directory
        at io.github.dreamylost.graphql.codegen.GraphQLCodegenPlugin.$anonfun$projectSettings$14(GraphQLCodegenPlugin.scala:225)
        at scala.Function1.$anonfun$compose$1(Function1.scala:49)
        at sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:62)
        at sbt.std.Transform$$anon$4.work(Transform.scala:68)
        at sbt.Execute.$anonfun$submit$2(Execute.scala:282)
        at sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:23)
        at sbt.Execute.work(Execute.scala:291)
        at sbt.Execute.$anonfun$submit$1(Execute.scala:282)
        at sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:265)
        at sbt.CompletionService$$anon$2.call(CompletionService.scala:64)

You mentioned this? `Maybe it's related to this line of code, so I'll delete it to avoid potential concurrency problems.`

But I guess you deleted this comment after?  Can this be related?

**Reply by jxnu-liguobin** [reply]
@nnbbz The `cleanFiles` in that line of code only listens to and deletes the directory during the clean task. So it's not sure if it does have an impact.Plugin will be delete files during `clean` task, while `graphqlCodegen` will be delete files before creation, so I don't know when `clean` and `graphqlCodegen` will be triggered during `compile` execution? Is it possible for two tasks to be concurrent? It's not easy to reproduce. Are you modifying the code while the play app is running and then reloading（will compile） it automatically? You can explain your steps in detail, which may help to reproduce the problem.

Note: Incremental compilation needs to be considered

**Reply by nnbbz** [reply]
@jxnu-liguobin 
`Are you modifying the code while the play app is running and then reloading（will compile） it automatically`
Yes I'm actually doing this.  When I run the app and make modifications to the code, it will automatically compile.  This generally leads to an exception where it fails to delete the directory and after that, it will not recover unless I restart sbt.

**Reply by jxnu-liguobin** [reply]
@nnbbz Do you use idea SBT shell or stand-alone SBT? What is the version?

**Reply by jxnu-liguobin** [reply]
I exec `sbt run` with idea or terminal or  sbt shell but it didn't reappear.
Are we missing some key information? Could you provide your `build.sbt`？

1. By idea
![image](https://user-images.githubusercontent.com/29496873/108810070-3f022d80-75e5-11eb-9d05-c7d0e22667d9.png)

2. By terminal
![image](https://user-images.githubusercontent.com/29496873/108810169-7e307e80-75e5-11eb-9bc5-9d2d2102b03e.png)

3. By sbt shell

![image](https://user-images.githubusercontent.com/29496873/108810345-e717f680-75e5-11eb-98bd-fe7fec9d2bac.png)

**Reply by nnbbz** [reply]
@jxnu-liguobin Are you on windows?  It might be a windows issue?  I had a friend try it as well, and he wasn't getting this issue either.

**Reply by jxnu-liguobin** [reply]
@nnbbz No, ![image](https://user-images.githubusercontent.com/29496873/109107189-4142c400-776c-11eb-8847-5dc8196e7235.png)
